libraries/stdlib/samples/test/samples/collections/collections.kt

```diff
@@ -23,14 +23,224 @@ import kotlin.test.*
+            val empty = listOfNotNull<Any>(null)
+            assertPrints(empty, "[]")
+
+            val list = listOfNotNull(1, null, 2, null, 3)
+            assertPrints(list, "[1, 2, 3]")
+        }
+
+        @Sample
+        fun readOnlyListFromInitializer() {
+            val empty = List(0) {}
+            assertPrints(empty, "[]")
+
+            val squares = List(5) { (it + 1) * (it + 1) }
+            assertPrints(squares, "[1, 4, 9, 16, 25]")
+        }
+
+        @Sample
+        fun mutableListFromInitializer() {
+            val chars = listOf('a', 'b', 'c', 'd')
+
+            val list = MutableList(3) { chars[it] }
+            assertPrints(list, "[a, b, c]")
+
+            list.add(chars[3])
+            assertTrue(list == chars)
+        }
+
         @Sample
         fun lastIndexOfList() {
             assertPrints(emptyList<Any>().lastIndex, "-1")
             val list = listOf("a", "x", "y")
             assertPrints(list.lastIndex, "2")
             assertPrints(list[list.lastIndex], "y")
         }
+
+        @Sample
+        fun listOrEmpty() {
+            val nullList: List<Any>? = null
+            assertPrints(nullList.orEmpty(), "[]")
+
+            val list: List<Char>? = listOf('a', 'b', 'c')
+            assertPrints(list.orEmpty(), "[a, b, c]")
+        }
+
+        @Sample
+        fun enumerationToList() {
+            val numbers = java.util.Hashtable<String, Int>()
+            numbers.put("one", 1)
+            numbers.put("two", 2)
+            numbers.put("three", 3)
+
+            val enumeration = numbers.elements()
+            val list = enumeration.toList().sorted()
+            assertPrints(list, "[1, 2, 3]")
+        }
+
+        @Sample
+        fun binarySearchOnComparable() {
+            val list = mutableListOf('a', 'b', 'c', 'd', 'e')
+            assertPrints(list.binarySearch('d'), "3")
+
+            list.remove('d')
+
+            val invertedInsertionPoint = list.binarySearch('d')
+            val actualInsertionPoint = -(invertedInsertionPoint + 1)
+            assertPrints(actualInsertionPoint, "3")
+
+            list.add(actualInsertionPoint, 'd')
+            assertPrints(list, "[a, b, c, d, e]")
+        }
+
+        @Sample
+        fun binarySearchWithComparator() {
+            val colors = listOf("Blue", "green", "ORANGE", "Red", "yellow")
+            assertPrints(colors.binarySearch("RED", String.CASE_INSENSITIVE_ORDER), "3")
+        }
+
+        @Sample
+        fun binarySearchBySelector() {
+            data class Box(val value: Int)
+
+            val numbers = listOf(1, 3, 7, 10, 12)
+            val boxes = numbers.map { Box(it) }
+            assertPrints(boxes.binarySearchBy(10) { it.value }, "3")
+        }
+
+        @Sample
+        fun binarySearchByKeyWithComparator() {
+            data class Box(val value: Int)
+            val comparator = Comparator<Box> { x, y -> y.value - x.value }
+
+            val numbers = listOf(12, 10, 7, 3, 1)
```

Comment:
It might be unclear from this example that the list must be sorted so that the sign of `comparison` is monotonically increasing.